### PR TITLE
expose port range

### PR DESCRIPTION
**Pull request by fahedouch**
Refacto `portutil` && fixing https://github.com/containerd/nerdctl/issues/163
Signed-off-by: fahed dorgaa <fahed.dorgaa@gmail.com>

### CODE REVIEW

File: pkg/portutil/portutil.go [hunk 1]
```diff
@@ -42,44 +59,55 @@ func ParseFlagP(s string) (*gocni.PortMapping, error) {
...
 		}
-		res.ContainerPort = int32(i)
-		return res, nil
-	default:
-		return nil, errors.Errorf("failed to parse %q, unexpected colons", s)
+
+		multi_res = append(multi_res, res)
 	}
+
+	return multi_res, nil
```
Comment (AkihiroSuda) [inline pkg/portutil/portutil.go:112]: underscore is discouraged in Go

File: pkg/portutil/portutil.go [hunk 2]
```diff
@@ -42,44 +59,55 @@ func ParseFlagP(s string) (*gocni.PortMapping, error) {
...
+	}
+
+	startPort, endPort, err := nat.ParsePortRange(containerPort)
+	if err != nil {
+		return nil, errors.Errorf("Invalid containerPort: %s", containerPort)
+	}
+
+	if hostPort != "" && (endPort-startPort) != (endHostPort-startHostPort) {
+		if endPort != startPort {
+			return nil, errors.Errorf("Invalid ranges specified for container and host Ports: %s and %s", containerPort, hostPort)
```
Comment (AkihiroSuda) [inline pkg/portutil/portutil.go:90]: Go discourages using an error string that start with a lower character

File: run_network_test.go [hunk 3]
```diff
@@ -146,46 +146,120 @@ func TestRunPort(t *testing.T) {
...
+			connectURLPort: 7000,
+		},
+		{
+			listenIP:       zeroIP,
+			connectIP:      lo,
+			hostPort:       "7000-7005",
+			containerPort:  "80",
+			connectURLPort: 7005,
+			err:            "connection refused",
 		},
```
Comment (AkihiroSuda) [inline run_network_test.go:263]: Can we also test an invalid range such as `-p 7000-7005:79-85`

File: pkg/portutil/portutil.go [hunk 4]
```diff
@@ -18,14 +18,31 @@ package portutil
...
-	"strconv"
 	"strings"
 
 	gocni "github.com/containerd/go-cni"
+	"github.com/docker/go-connections/nat"
 	"github.com/pkg/errors"
 )
 
-func ParseFlagP(s string) (*gocni.PortMapping, error) {
+func splitParts(rawport string) (string, string, string) {
```
Comment (AkihiroSuda) [inline pkg/portutil/portutil.go:28]: needs comment lines to explain the ret values

File: pkg/portutil/portutil.go [hunk 5]
```diff
@@ -18,14 +18,31 @@ package portutil
...
+
+	switch n {
+	case 1:
+		return "", "", containerport
+	case 2:
+		return "", parts[0], containerport
+	case 3:
+		return parts[0], parts[1], containerport
+	default:
+		return strings.Join(parts[:n-2], ":"), parts[n-2], containerport
```
Comment (AkihiroSuda) [inline pkg/portutil/portutil.go:41]: The switch is hard to read. Is this `default` case for IPv6?

File: pkg/portutil/portutil.go [hunk 6]
```diff
@@ -42,44 +59,55 @@ func ParseFlagP(s string) (*gocni.PortMapping, error) {
...
 
-	splitByColon := strings.Split(splitBySlash[0], ":")
-	switch len(splitByColon) {
-	case 1:
+	multi_res := []gocni.PortMapping{}
+
+	ip, hostPort, containerPort := splitParts(splitBySlash[0])
+
+	if containerPort == "" {
+		return nil, errors.Errorf("No port specified: %s<empty>", splitBySlash[0])
```
Comment (AkihiroSuda) [inline pkg/portutil/portutil.go:71]: What does `<empty>` mean?

File: pkg/portutil/portutil.go (hunk 5 above)
Comment (fahedouch) [reply pkg/portutil/portutil.go:41]: Nope, I targeting Ipv4 here

look a this two example :
`./nerdctl --snapshotter native run -d -p 1515:127.0.0.1:80:80 alpine sleep 100` 

parts = [1515 127.0.0.1 80 80]

parts[n-1] -> 80
parts[n-2]  -> 80
parts[:n-2] -> [ 1515 127.0.0.1]  So strings.Join(parts[:n-2], ":") => 1515:127.0.0.1 (original input)

`./nerdctl --snapshotter native run -d -p 1515-127.0.0.1:80:80 alpine sleep 100` 

parts = [1515-127.0.0.1 80 80]

parts[n-1] -> 80
parts[n-2]  -> 80
parts[:n-2] -> [1515-127.0.0.1]  So strings.Join(parts[:n-2], ":") => 1515-127.0.0.1 

it is ok for ipv6 too.

File: pkg/portutil/portutil.go (hunk 6 above)
Comment (fahedouch) [reply pkg/portutil/portutil.go:71]: nothing

File: pkg/portutil/portutil.go (hunk 4 above)
Comment (fahedouch) [reply pkg/portutil/portutil.go:30]: done

File: run_network_test.go (hunk 3 above)
Comment (fahedouch) [reply run_network_test.go:287]: done

File: pkg/portutil/portutil.go (hunk 5 above)
Comment (AkihiroSuda) [reply pkg/portutil/portutil.go:42]: I don't understand what `1515-127.0.0.1:80:80` means. Is it officially supported by Docker?

File: run_network_test.go [hunk 7]
```diff
@@ -146,48 +146,147 @@ func TestRunPort(t *testing.T) {
...
-		connectIP net.IP
-		port      int
-		err       string
+		listenIP       net.IP
+		connectIP      net.IP
+		hostPort       string
+		containerPort  string
+		connectURLPort int
+		assert         string
+		err            string
```
Comment (AkihiroSuda) [inline run_network_test.go:155]: `assert string` should be probably `runExitCode int` or `runShouldSuccess bool`

File: run_network_test.go [hunk 8]
```diff
@@ -196,9 +295,16 @@ func TestRunPort(t *testing.T) {
 			testContainerName := fmt.Sprintf("nerdctl-test-nginx-%d", i)
 			base := testutil.NewBase(t)
 			defer base.Cmd("rm", "-f", testContainerName).Run()
-			pFlag := fmt.Sprintf("%s:%d:80", tc.listenIP.String(), tc.port)
-			connectURL := fmt.Sprintf("http://%s:%d", tc.connectIP.String(), tc.port)
+			pFlag := fmt.Sprintf("%s:%s:%s", tc.listenIP.String(), tc.hostPort, tc.containerPort)
+			connectURL := fmt.Sprintf("http://%s:%d", tc.connectIP.String(), tc.connectURLPort)
 			t.Logf("pFlag=%q, connectURL=%q", pFlag, connectURL)
+			if tc.assert != "ok" {
```
Comment (AkihiroSuda) [inline run_network_test.go:301]: This check should occur after creating `base.Cmd("run", "-d", ...)`

File: pkg/portutil/portutil.go (hunk 5 above)
Comment (fahedouch) [reply pkg/portutil/portutil.go:42]: @AkihiroSuda  it is a wrong IP that a user may try (just an example)

the purpose is to get  the IP block and check it.

File: run_network_test.go (hunk 8 above)
Comment (fahedouch) [reply run_network_test.go]: done

File: run_network_test.go (hunk 7 above)
Comment (fahedouch) [reply run_network_test.go]: done

**Comment by fahedouch** [comment]
@AkihiroSuda  can you re-review plz.

File: run_network_test.go [hunk 9]
```diff
@@ -196,13 +295,22 @@ func TestRunPort(t *testing.T) {
...
+					"--name", testContainerName,
+					"-p", pFlag,
+					testutil.NginxAlpineImage).AssertOK()
+			} else {
+				base.Cmd("run", "-d",
+					"--name", testContainerName,
+					"-p", pFlag,
+					testutil.NginxAlpineImage).AssertFail()
+				return
+			}
```
Comment (AkihiroSuda) [inline run_network_test.go:313]:
```go
cmd := base.Cmd("run", "-d", "--name", testContainerName, "-p", pFlag, testutil.NginxAlpineImage)
if tc.runShouldSuccess {
  cmd.AssertOK()
} else {
  cmd.AssertFail()
  return
}

Comment (fahedouch) [reply run_network_test.go:313]: Ah ok. I fix this.

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**